File: trees/OrdinaryTree/DFS/DeepFirstSearch.js
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  insert(value) {
    let node = new Node(value);
    if (this.root === null) {
      this.root = node;
      return this;
    }
    let current = this.root;
    if (current.value === value) return this;
    while (true) {
      // Left branch
      if (current.value > value) {
        if (current.left === null) {
          current.left = node;
          return this;
        }
        current = current.left;
      }
      // Right branch
      else if (current.value < value) {
        if (current.right === null) {
          current.right = node;
          return this;
        }
        current = current.right;
      }
    }
  }

  find(value) {
    if (this.root === null) return false;
    let current = this.root;
    while (current) {
      if (current.value === value) return current;
      if (current.value > value) {
        current = current.left;
      } else if (current.value < value) {
        current = current.right;
      }
    }
    return false;
  }

  contains(value) {
    if (this.root === null) return false;
    let current = this.root;
    while (current) {
      if (current.value === value) return true;
      if (current.value > value) {
        current = current.left;
      } else if (current.value < value) {
        current = current.right;
      }
    }
    return false;
  }

  DFSPreOrder() {
    let visited = [];

    function traverse(node) {
      visited.push(node.value);
      if (node.left) traverse(node.left);
      if (node.right) traverse(node.right);
    }
    traverse(this.root);
    return visited;
  }

  DFSPostOrder() {
    let visited = [];

    function traverse(node) {
      if (node.left) traverse(node.left);
      if (node.right) traverse(node.right);
      visited.push(node.value);
    }
    traverse(this.root);
    return visited;
  }

  DFSInOrder() {
    let visited = [];

    function traverse(node) {
      if (node.left) traverse(node.left);
      visited.push(node.value);
      if (node.right) traverse(node.right);
    }
    traverse(this.root);
    return visited;
  }
}


bs = new BinarySearchTree();
bs.insert(10);
bs.insert(6);
bs.insert(15);
bs.insert(3);
bs.insert(8);
bs.insert(20);
console.log(bs.DFSInOrder());

File: trees/OrdinaryTree/DFS/deep_first_search.py
import queue


class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def __str__(self):
        return f'< {self.value} >'

    def __repr__(self):
        return f'< {self.value} >'


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        node = Node(value)
        if not self.root:
            self.root = node
            return self

        current = self.root
        while True:
            # LEFT
            if current.value > value:
                if not current.left:
                    current.left = node
                    return self
                current = current.left

            # RIGHT
            if current.value < value:
                if not current.right:
                    current.right = node
                    return self
                current = current.right

    def find(self, value):
        current = self.root

        while True:
            if not current:
                return False
            if current.value == value:
                return current

            if current.value > value:
                current = current.left
            elif current.value < value:
                current = current.right

    def DFS_pre_order(self, current=None, visited=[]):
        if not self.root in visited:
            current = self.root

        if current:
            visited.append(current)
            if current.left: self.DFS_pre_order(current.left, visited)
            if current.right: self.DFS_pre_order(current.right, visited)

        return visited


bst = BinarySearchTree()
bst.insert(12)
bst.insert(10)
bst.insert(14)
bst.insert(23)
bst.insert(1)
bst.insert(15)
bst.insert(5)
bst.insert(11)
bst.insert(13)
print(bst.DFS_pre_order())
